spinup_bis/algos/tf2/bsp/core.py

```diff
@@ -0,0 +1,226 @@
+        activation (tf.function): A hidden layers activations specification.
+        act_noise (float): Stddev for Gaussian exploration noise.
+        ac_number (int): Number of the actor-critic models in the ensemble.
+        prior_weight (float): Randomly initialized network output scaling.
+    """
+
+    def __init__(self, observation_space, action_space, hidden_sizes,
+                 activation, prior_weight, act_noise, ac_number):
+        self._obs_dim = observation_space.shape[0]
+        self._act_dim = action_space.shape[0]
+        self._act_scale = action_space.high[0]
+        self._hidden_sizes = hidden_sizes
+        self._activation = activation
+        self._act_noise_dumping = act_noise[0]
+        self._act_noise_stddev = act_noise[1]
+        self._ac_number = ac_number
+        self._prior_weight = prior_weight
+
+    def _make_actor(self):
+        """Constructs and returns the actor model (tf.keras.Model)."""
+        obs_input = tf.keras.Input(shape=(self._obs_dim,))
+        body = mlp(self._hidden_sizes, self._activation)(obs_input)
+        mu = tf.keras.layers.Dense(self._act_dim)(body)
+
+        # Normalize the actions.
+        g = tf.reduce_mean(tf.math.abs(mu), axis=1, keepdims=True)
+        g = tf.maximum(g, tf.ones_like(g))
+        mu = mu / g
+
+        # Add the action noise.
+        pi = OUProcess(self._act_dim, self._act_noise_dumping,
+                       self._act_noise_stddev)(mu)
+
+        # Put the actions in the limit.
+        mu = tf.tanh(mu) * self._act_scale
+        pi = tf.tanh(pi) * self._act_scale
+
+        return tf.keras.Model(inputs=obs_input, outputs=[mu, pi])
+
+    def make_actor(self):
+        """Constructs and returns the ensemble of actor models."""
+        obs_inputs = tf.keras.Input(shape=(None, self._obs_dim),
+                                    batch_size=self._ac_number)
+        mus, pis = [], []
+        for obs_input in tf.unstack(obs_inputs, axis=0):
+            model = self._make_actor()
+            mu, pi = model(obs_input)
+            mus.append(mu)
+            pis.append(pi)
+        return tf.keras.Model(inputs=obs_inputs, outputs=[
+            tf.stack(mus, axis=0),
+            tf.stack(pis, axis=0),
+        ])
+
+    def _make_critic(self, trainable=True):
+        """Constructs and returns the critic model (tf.keras.Model)."""
+        obs_input = tf.keras.Input(shape=(self._obs_dim,))
+        act_input = tf.keras.Input(shape=(self._act_dim,))
+
+        concat_input = tf.keras.layers.Concatenate(
+            axis=-1)([obs_input, act_input])
+
+        q = tf.keras.Sequential([
+            mlp(self._hidden_sizes, self._activation, trainable=trainable),
+            tf.keras.layers.Dense(1, trainable=trainable),
+            tf.keras.layers.Reshape([])  # Very important to squeeze values!
+        ])(concat_input)
+
+        return tf.keras.Model(inputs=[obs_input, act_input], outputs=q)
+
+    def make_critic(self):
```

Comment:
Remove this (ensemble) and then remove `_` from the factory method above.

Reply:
done